plugin/symbols.py

```diff
@@ -42,53 +41,118 @@ def format_symbol_kind(kind: int) -> str:
+        self.regions = []  # type: List[Tuple[sublime.Region, Optional[sublime.Region]]]
+        self.found_at_least_one_nonempty_detail = False
+        self.is_first_selection = False
 
     def is_enabled(self, event: Optional[dict] = None) -> bool:
         return self.has_client_with_capability('documentSymbolProvider')
 
     def run(self, edit: sublime.Edit) -> None:
         client = self.client_with_capability('documentSymbolProvider')
-        file_path = self.view.file_name()
-        if client and file_path:
-            params = {
-                "textDocument": {
-                    "uri": filename_to_uri(file_path)
-                }
-            }
-            request = Request.documentSymbols(params)
-            client.send_request(request, self.handle_response)
-
-    def handle_response(self, response: Optional[List]) -> None:
-        response_list = response or []
-        symbols = list(format_symbol(item) for item in response_list)
-        self.symbols = response_list
+        if client:
+            client.send_request(
+                Request.documentSymbols({"textDocument": text_document_identifier(self.view)}), self.handle_response)
+
+    def handle_response(self, response: Any) -> None:
         window = self.view.window()
-        if window:
-            window.show_quick_panel(symbols, self.on_symbol_selected)
+        if window and isinstance(response, list) and len(response) > 0:
+            self.old_regions = [sublime.Region(r.a, r.b) for r in self.view.sel()]
+            self.is_first_selection = True
+            window.show_quick_panel(
+                self.process_symbols(response),
+                self.on_symbol_selected,
+                sublime.KEEP_OPEN_ON_FOCUS_LOST,
+                0,
+                self.on_highlighted)
+            self.view.run_command("lsp_selection_clear")
 
-    def on_symbol_selected(self, symbol_index: int) -> None:
-        if symbol_index == -1:
-            return
-        selected_symbol = self.symbols[symbol_index]
-        range = selected_symbol.get('location', selected_symbol.get('range'))
-        range = range.get('range', range)
-        if not range:
-            debug('could not recognize the type: expected either SymbolInformation or DocumentSymbol')
+    def region(self, index: int) -> sublime.Region:
+        return self.regions[index][0]
+
+    def selection_region(self, index: int) -> Optional[sublime.Region]:
+        return self.regions[index][1]
+
+    def on_symbol_selected(self, index: int) -> None:
+        if index == -1:
+            if len(self.old_regions) > 0:
+                self.view.run_command("lsp_selection_add", {"regions": [(r.a, r.b) for r in self.old_regions]})
+                self.view.show_at_center(self.old_regions[0].begin())
+        else:
+            region = self.selection_region(index) or self.region(index)
+            self.view.run_command("lsp_selection_add", {"regions": [(region.a, region.a)]})
```

Comment:
Just looking at this file here... is this a typo and should be `(region.a, region.b)`?

Reply:
It's a matter of preference. With (region.a, region.b) it will select the whole entity. While the built-in Goto Symbol indeed does that, the behavior is weird for this feature. Because the region can be large (e.g. a body of a large function, or a namespace). This is also the region why `add_regions` and `erase_regions` is used instead of "highlighting" entities with `view.sel()`. The regions for the built-in Goto Symbol never span more than 1 line.

Reply:
All fine then, thanks for explanation.